File: nodemanager/core/JobTaskTable.cpp
#include "JobTaskTable.h"
#include "../utils/WriterLock.h"
#include "../utils/ReaderLock.h"
#include "../utils/System.h"
#include <math.h>

using namespace hpc::core;
using namespace web;
using namespace hpc::data;
using namespace hpc::utils;

JobTaskTable* JobTaskTable::instance = nullptr;

json::value JobTaskTable::ToJson()
{
    ReaderLock readerLock(&this->lock);
    auto j = this->nodeInfo.ToJson();
    return std::move(j);
}

int JobTaskTable::GetTaskCount()
{
    ReaderLock readerLock(&this->lock);
    int taskCount = 0;

    for_each(this->nodeInfo.Jobs.begin(), this->nodeInfo.Jobs.end(),
        [&taskCount] (auto& i) { taskCount += i.second->Tasks.size(); });

    return taskCount;
}

int JobTaskTable::GetCoresInUse()
{
    ReaderLock readerLock(&this->lock);

    bool allUsed = false;
    int cores, sockets;
    System::CPU(cores, sockets);

    std::vector<uint64_t> coresMask(ceil((float)cores / 64), 0);
    for_each(this->nodeInfo.Jobs.begin(), this->nodeInfo.Jobs.end(), [&coresMask, &allUsed] (auto& i)
    {
        for_each(i.second->Tasks.begin(), i.second->Tasks.end(), [&coresMask, &allUsed] (auto& t)
        {
            if (t.second->Affinity.empty())
            {
                allUsed = true;
            }
            else
            {
                for (size_t i = 0; i < coresMask.size(); i++)
                {
                    coresMask[i] |= t.second->Affinity[i];
                }
            }
        });
    });

    if (allUsed)
    {
        return cores;
    }

    int used = 0;
    uint64_t bit = 1;
    for (int i = 0; i < cores; i++)
    {
        if (coresMask[i / 64] & (bit << i % 64))
        {
            used++;
        }
    }

    return used;
}

std::shared_ptr<TaskInfo> JobTaskTable::AddJobAndTask(int jobId, int taskId, bool& isNewEntry)
{
    WriterLock writerLock(&this->lock);

    std::shared_ptr<JobInfo> job;
    auto j = this->nodeInfo.Jobs.find(jobId);
    if (j == this->nodeInfo.Jobs.end())
    {
        job = std::shared_ptr<JobInfo>(new JobInfo(jobId));
        this->nodeInfo.Jobs[jobId] = job;
    }
    else
    {
        job = j->second;
    }

    std::shared_ptr<TaskInfo> task;
    auto t = job->Tasks.find(taskId);
    if (t == job->Tasks.end())
    {
        task = std::shared_ptr<TaskInfo>(new TaskInfo(jobId, taskId, nodeInfo.Name));
        job->Tasks[taskId] = task;
        isNewEntry = true;
    }
    else
    {
        task = t->second;
        isNewEntry = false;
    }

    return task;
}

std::shared_ptr<TaskInfo> JobTaskTable::GetTask(int jobId, int taskId)
{
    ReaderLock readerLock(&this->lock);

    std::shared_ptr<TaskInfo> task;

    auto j = this->nodeInfo.Jobs.find(jobId);
    if (j != this->nodeInfo.Jobs.end())
    {
        auto t = j->second->Tasks.find(taskId);
        if (t != j->second->Tasks.end())
        {
            task = t->second;
        }
    }

    return task;
}

std::shared_ptr<JobInfo> JobTaskTable::RemoveJob(int jobId)
{
    WriterLock writerLock(&this->lock);

    std::shared_ptr<JobInfo> job;
    auto j = this->nodeInfo.Jobs.find(jobId);
    if (j != this->nodeInfo.Jobs.end())
    {
        job = j->second;
        this->nodeInfo.Jobs.erase(j);
    }

    return job;
}

void JobTaskTable::RemoveTask(int jobId, int taskId, uint64_t attemptId)
{
    WriterLock writerLock(&this->lock);

    auto j = this->nodeInfo.Jobs.find(jobId);
    if (j != this->nodeInfo.Jobs.end())
    {
        auto t = j->second->Tasks.find(taskId);

        // only erase when attempt ID matches.
        if (t != j->second->Tasks.end() &&
            t->second->GetAttemptId() == attemptId)
        {
            j->second->Tasks.erase(t);
        }
    }
}

File: src/NodeAgent/Utils/JobTaskTable.cs
using NodeAgent.Models;

namespace NodeAgent.Utils;


//The public methods are not thread-safe. The caller has to ensure the thread-safety.
public class JobTaskTable
{
    private IDictionary<int, JobInfo> _jobs = new Dictionary<int, JobInfo>();

    //Add jobs task if it's not existed in _jobs. Return the existed or new TaskInfo.
    public TaskInfo AddJobAndTask(int jobId, int taskId, out bool isNewEntry)
    {
        if (!_jobs.TryGetValue(jobId, out var jobInfo))
        {
            jobInfo = new JobInfo() { JobId = jobId };
            _jobs.Add(jobId, jobInfo);
        }

        if (jobInfo.Tasks.TryGetValue(taskId, out var taskInfo))
        {
            isNewEntry = false;
        }
        else
        {
            isNewEntry = true;
            taskInfo = new TaskInfo() { JobId = jobId, TaskId = taskId };
            jobInfo.Tasks.Add(taskId, taskInfo);
        }

        return taskInfo;
    }

    public TaskInfo? GetTask(int jobId, int taskId)
    {
        if (_jobs.TryGetValue(jobId, out var jobInfo)
            && jobInfo.Tasks.TryGetValue(taskId, out var taskInfo))
        {
            return taskInfo;
        }
        return null;
    }

    public TaskInfo? RemoveTask(int jobId, int taskId, ulong attemptId)
    {
        TaskInfo? taskInfo = null;
        if (_jobs.TryGetValue(jobId, out var jobInfo))
        {
            if (jobInfo.Tasks.TryGetValue(taskId, out taskInfo) && taskInfo.AttemptId == attemptId)
            {
                jobInfo.Tasks.Remove(taskId);
            }
        }
        return taskInfo;
    }

    public JobInfo? RemoveJob(int jobId)
    {
        if (_jobs.TryGetValue(jobId, out var jobInfo))
        {
            _jobs.Remove(jobId);
        }
        return jobInfo;
    }

    public int GetJobCount()
    {
        return _jobs.Count;
    }

    public int GetTaskCount()
    {
        return GetTasks().Count();
    }

    public int GetCoresInUse()
    {
        throw new NotImplementedException();
    }

    public IEnumerable<TaskInfo> GetTasks()
    {
        foreach (var job in _jobs.Values)
        {
            if (job.Tasks == null)
            {
                continue;
            }
            foreach (var task in job.Tasks.Values)
            {
                yield return task;
            }
        }
    }

    public IEnumerable<JobInfo> GetJobs(bool copy = false)
    {
        foreach (var job in _jobs.Values)
        {
            yield return copy ? job.Copy() : job;
        }
    }
}
